storage/innobase/row/row0merge.cc

```diff
@@ -1166,16 +1159,21 @@ dberr_t row_merge_buf_write(const row_merge_buf_t *buf,
 
 		if (blob_file) {
 			ut_ad(buf->index->is_primary());
-			err = row_merge_buf_blob(
-				entry, n_fields, &blob_heap, blob_file);
+			if (!blob_heap) {
+				blob_heap = mem_heap_create(100);
+			}
+
+			err = row_merge_buf_blob(entry->fields, n_fields,
+						 blob_file, buf->index,
+                                                 &blob_heap);
 			if (err != DB_SUCCESS) {
 				goto func_exit;
 			}
 		}
 
 		ulint rec_size= row_merge_buf_encode(
 				&b, index, entry, n_fields);
-		if (blob_file && rec_size > srv_page_size) {
+		if (blob_file && rec_size > BTR_PAGE_MAX_REC_SIZE) {
 			err = DB_TOO_BIG_RECORD;
 			goto func_exit;
 		}
```

Comment:
This constant looks bogus to me and had better be replaced (in a separate bug ticket, targeting 10.5). Isn’t the maximum size of a record limited to 16383 bytes? In `ROW_FORMAT=REDUNDANT`, we reserve 7 or 14 bits per index field to store the end offset of each field. The most significant bit is the NULL flag, and in case of 2-byte offsets we have another flag `REC_2BYTE_EXTERN_MASK=1<<14`. In other formats, the limit is 16383 bytes per record field, for variable-length fields. See 9284e8b2c64aba1ac108512ecc415a2ea563c387.